File: src/service/bookService.js
import {getRequest, postRequest} from "../utils/ajax";
import {constants} from "../config/constants";

export function getBooksByPage(page, callback) {
    const url = `${constants.base_url}book?page=${page}`;
    getRequest(url, callback);
}

export function searchBooksByTags(tags, callback) {
    let tags_str = '';
    for (let i = 0; i < tags.length; ++i) {
        tags_str = tags_str.concat(tags[i] + ',');
    }
    tags_str = tags_str.substr(0, tags_str.length - 1);
    console.log("tags: " + tags_str);
    const url = `${constants.base_url}book/tag/${tags_str}`;

    getRequest(url, callback);
}

export function getBooks(callback) {
    const url = `${constants.base_url}book/getAll`;
    getRequest(url, callback);
}

export function getBooksByKeyword(keyword, callback) {
    const url = `${constants.base_url}book/search?keyword=${keyword}`;
    getRequest(url, callback);
}

export function getBookById(bookId, callback) {
    const url = `${constants.base_url}book/get?bookId=${Number(bookId)}`;
    getRequest(url, callback);
}

export function getConcernedBookInfo(bookTitle, website, callback) {
    const url = `${constants.base_url}book/concerned?bookTitle=${bookTitle}&website=${website}`;
    getRequest(url, callback);
}

export function deleteBooks(bookIdList, callback) {
    const url = `${constants.base_url}book/delete`;
    postRequest(url, bookIdList, callback)
}

export function undercarriageBooks(bookIdList, callback) {
    const url = `${constants.base_url}book/undercarriage`;
    postRequest(url, bookIdList, callback)
}

export function putOnSale(bookIdList, callback) {
    const url = `${constants.base_url}book/putOnSale`;
    postRequest(url, bookIdList, callback)
}

export function postModifiedBook(book, callback) {
    const url = `${constants.base_url}book/modify`;
    postRequest(url, book, callback);
}

export function getRankedBooks(startNEndDates, callback) {
    console.log("startNEndDates", startNEndDates);
    if (startNEndDates.length === 2) {
        const url = `${constants.base_url}book/rank?from=${startNEndDates[0]}&to=${startNEndDates[1]}`;
        getRequest(url, callback)
    } else {
        const url = `${constants.base_url}book/rank/all`;
        getRequest(url, callback);
    }
}

File: backend/src/main/java/com/catstore/service/BookService.java
package com.catstore.service;

import com.catstore.entity.Book;
import com.catstore.entity.Book4Neo;
import com.catstore.entity.BookTag;
import com.catstore.model.Message;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Map;

public interface BookService {
    void saveBookAndTags(Book4Neo book, List<BookTag> tags);

    List<Book> getBooks();

    Message getBooks(int page);

    Book getBookById(Integer bookId);

    List<Book> getBooksByKeyword(String keyword);

    String getAuthorByTitle(String title);

    String getBookTitleByBookId(Integer bookId);

    List<Map<String, String>> getConcernedBookInfo(String bookTitle, String websiteSrc);

    Boolean deleteBooks(ArrayList<Integer> bookIdList);

    Boolean undercarriage(ArrayList<Integer> bookIdList);

    Boolean putOnSale(ArrayList<Integer> bookIdList);

    ArrayList<Book> getRankedBooks(Date from, Date to);

    ArrayList<Book> getAllRankedBooks();

    void postModifiedBook(Map<String, String> book);

    List<Book> searchByTags(List<String> tags);
}
